### PR TITLE
Implementation of fetch prop

### LINKED ISSUE #565: [nGQL] Get properties from vertex/edge.
Here I propose a syntax about getting properties from vertex/edge:
```
fetch vertex on [_id:]<vertex_id> [yield ...]
```

```
fetch edge on [(_src,_dst,_type,_rank):](<src_vid>, <dst_vid>, <edge_type>, <rank>) [yield ...]
```

@dutor @sherman-the-tank @whitewum

### PR DESCRIPTION
**Pull request by CPWstatic**
close #565

### CODE REVIEW

File: src/graph/test/FetchEdgesTest.cpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    }
+};
+
+TEST_F(FetchEdgesTest, base) {
+    {
+        cpp2::ExecutionResponse resp;
+        auto &player = players_["Boris Diaw"];
+        auto &serve = player.serves()[0];
+        auto &team = teams_[std::get<0>(serve)];
+        auto *fmt = "FETCH PROP ON serve %ld->%ld"
```
Comment (whitewum) [inline src/graph/test/FetchEdgesTest.cpp]: why call `FETCH PROP ON`, i think  `FETCH` is enough

BTW, what about the sql-like sentence: `(better)YIELD xxx ON serve`
or (worse) `ON serve YIELD`.

File: src/graph/test/FetchEdgesTest.cpp [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        auto code = client_->execute(query, resp);
+        ASSERT_EQ(cpp2::ErrorCode::SUCCEEDED, code);
+        std::vector<std::tuple<int64_t, int64_t>> expected = {
+            {std::get<1>(serve), std::get<2>(serve)},
+        };
+        ASSERT_TRUE(verifyResult(resp, expected));
+    }
+}
+
+TEST_F(FetchEdgesTest, noYield) {
```
Comment (whitewum) [inline src/graph/test/FetchEdgesTest.cpp]: great

File: src/graph/test/FetchEdgesTest.cpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+TEST_F(FetchEdgesTest, nonExistEdge) {
+    {
+        cpp2::ExecutionResponse resp;
+        auto query = "FETCH PROP ON serve hash(\"Zion Williamson\")->hash(\"Spurs\")"
+                     " YIELD serve.start_year";
+        auto code = client_->execute(query, resp);
+        ASSERT_EQ(cpp2::ErrorCode::SUCCEEDED, code);
+        ASSERT_EQ(nullptr, resp.get_rows());
+    }
+}
```
Comment (whitewum) [inline src/graph/test/FetchEdgesTest.cpp]: let's test with distinct

fetch on 101,102 yield num | dedup

fetch on 101,102 yield distinct same_num_prop // should dedup or syntax error?

File: src/graph/test/FetchVerticesTest.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        std::vector<std::tuple<std::string, int64_t>> expected = {
+            {"Tony Parker", players_["Tony Parker"].age()},
+            {"Tim Duncan", players_["Tim Duncan"].age()},
+        };
+        ASSERT_TRUE(verifyResult(resp, expected));
+    }
+    {
+        cpp2::ExecutionResponse resp;
+        auto &player = players_["Boris Diaw"];
+        auto *fmt = "$var = GO FROM %ld over like;"
```
Comment (whitewum) [inline src/graph/test/FetchVerticesTest.cpp]: great

File: src/graph/test/FetchVerticesTest.cpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                    "FETCH PROP ON player $var.id YIELD player.name as name, player.age"
+                    " | ORDER BY name";
```
Comment (whitewum) [inline src/graph/test/FetchVerticesTest.cpp]: great

File: src/graph/FetchEdgesExecutor.cpp [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+            srcid_ = edgeKeyRef->srcid();
+            DCHECK_NOTNULL(srcid_);
+
+            dstid_ = edgeKeyRef->dstid();
+            DCHECK_NOTNULL(dstid_);
+
+            rank_ = edgeKeyRef->rank();
+            auto ret = edgeKeyRef->varname();
+            if (!ret.ok()) {
+                status = std::move(ret).status();
```
Comment (wadeliuyi) [inline src/graph/FetchEdgesExecutor.cpp]: why not break?

File: src/graph/FetchEdgesExecutor.cpp [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+            setupColumns();
+        } else {
+            yields_ = clause->columns();
+        }
+
+        for (auto *col : yields_) {
+            col->expr()->setContext(expCtx_.get());
+            status = col->expr()->prepare();
+            if (!status.ok()) {
+                break;
```
Comment (wadeliuyi) [inline src/graph/FetchEdgesExecutor.cpp]: why not return directly?

File: src/graph/FetchEdgesExecutor.cpp [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+    std::vector<int64_t> ranks;
+    if (rank_ != nullptr) {
+        ret = inputs->getVIDs(*rank_);
+        if (!ret.ok()) {
+            return ret.status();
+        }
+        ranks = std::move(ret).value();
+    }
+
+    for (auto index = 0u; index < srcVids.size(); ++index) {
```
Comment (wadeliuyi) [inline src/graph/FetchEdgesExecutor.cpp]: maybe deltype(srcVids.size()) is better than auto.

File: src/graph/FetchEdgesExecutor.cpp [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+        storage::cpp2::EdgeKey key;
+        key.set_src(srcVids[index]);
+        key.set_edge_type(edgeType_);
+        key.set_dst(dstVids[index]);
```
Comment (wadeliuyi) [inline src/graph/FetchEdgesExecutor.cpp]: Is dstVids.size() always same with srcVids.size()?

File: src/graph/FetchEdgesExecutor.cpp [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+    std::move(future).via(runner).thenValue(cb).thenError(error);
+}
+
+StatusOr<std::vector<storage::cpp2::PropDef>> FetchEdgesExecutor::getPropNames() {
+    std::vector<storage::cpp2::PropDef> props;
+    for (auto &prop : expCtx_->aliasProps()) {
+        storage::cpp2::PropDef pd;
+        pd.owner = storage::cpp2::PropOwner::EDGE;
+        pd.name = prop.second;
+        props.emplace_back(pd);
```
Comment (wadeliuyi) [inline src/graph/FetchEdgesExecutor.cpp]: emplace_back(storage::cpp2::PropOwner::EDGE, prop.second) more better?

File: src/graph/FetchEdgesExecutor.cpp [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+    auto all = result.responses();
+    std::shared_ptr<SchemaWriter> outputSchema;
+    std::unique_ptr<RowSetWriter> rsWriter;
+    for (auto &resp : all) {
+        if (!resp.__isset.schema || !resp.__isset.data) {
+            continue;
+        }
+
+        std::shared_ptr<ResultSchemaProvider> eschema;
+        if (resp.get_schema() != nullptr) {
```
Comment (wadeliuyi) [inline src/graph/FetchEdgesExecutor.cpp]: why not move the check to line 296 or define a check function?

File: src/graph/FetchEdgesExecutor.cpp [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+            eschema = std::make_shared<ResultSchemaProvider>(resp.schema);
+        } else {
+            continue;
+        }
+
+        auto *data = resp.get_data();
+        if (data == nullptr || data->empty()) {
```
Comment (wadeliuyi) [inline src/graph/FetchEdgesExecutor.cpp]: why not move the check to line 296 or define a check function?

File: src/graph/FetchEdgesExecutor.cpp [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+            continue;
+        }
+        RowSetReader rsReader(eschema, *data);
+        auto iter = rsReader.begin();
+        while (iter) {
+            if (outputSchema == nullptr) {
```
Comment (wadeliuyi) [inline src/graph/FetchEdgesExecutor.cpp]: maybe there are some method to move the check out of this loop.

File: src/graph/FetchEdgesExecutor.cpp [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+            auto collector = std::make_unique<Collector>(eschema.get());
+            auto writer = std::make_unique<RowWriter>(outputSchema);
+
+            auto &getters = expCtx_->getters();
+            getters.getAliasProp = [&] (const std::string&, const std::string &prop) {
+                return collector->collect(prop, &*iter, writer.get());
+            };
+            for (auto *column : yields_) {
+                auto *expr = column->expr();
+                auto value = expr->eval();
```
Comment (wadeliuyi) [inline src/graph/FetchEdgesExecutor.cpp]: what is the role about value?

File: src/graph/FetchEdgesExecutor.cpp (hunk 9 above)
Comment (CPWstatic) [reply src/graph/FetchEdgesExecutor.cpp]: Yes, I made some restrictions. They would come from same variable or input.

File: src/graph/FetchExecutor.cpp [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+        auto *expr = column->expr();
+        auto value = expr->eval();
+        record.emplace_back(std::move(value));
+    }
+
+    using nebula::cpp2::SupportedType;
+    for (auto index = 0u; index < record.size(); ++index) {
+        SupportedType type;
+        switch (record[index].which()) {
+            case 0:
```
Comment (wadeliuyi) [inline src/graph/FetchExecutor.cpp]: why not use micro?

File: src/graph/FetchVerticesExecutor.cpp [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+            } else if (expr->isVariableExpression()) {
+                auto *vexpr = static_cast<VariablePropertyExpression*>(expr);
+                varname_ = vexpr->var();
+                colname_ = vexpr->prop();
+            } else {
+                LOG(FATAL) << "Unknown kind of expression.";
+            }
+            break;
+        }
+    } while (false);
```
Comment (wadeliuyi) [inline src/graph/FetchVerticesExecutor.cpp]: I think the do {} while (false) is useless.

File: src/graph/FetchVerticesExecutor.cpp [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+    while (iter) {
+        auto *name = iter->getName();
+        auto *ref = new std::string("");
+        auto *tagName = new std::string(*tag);
+        Expression *expr = new AliasPropertyExpression(ref, tagName, new std::string(name));
+        YieldColumn *column = new YieldColumn(expr);
+        yieldColsHolder_->addColumn(column);
+        yields_.emplace_back(column);
+        ++iter;
+    }
```
Comment (wadeliuyi) [inline src/graph/FetchVerticesExecutor.cpp]: is the code in this loop same with function "FetchEdgesExecutor::setupColumns"?

File: src/graph/FetchVerticesExecutor.cpp [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+            vschema = std::make_shared<ResultSchemaProvider>(resp.vertex_schema);
+        }
+        for (auto &vdata : resp.vertices) {
+            std::unique_ptr<RowReader> vreader;
+            if (vschema != nullptr) {
+                if (!vdata.__isset.vertex_data || vdata.vertex_data.empty()) {
+                    continue;
+                }
+                vreader = RowReader::getRowReader(vdata.vertex_data, vschema);
+            }
```
Comment (wadeliuyi) [inline src/graph/FetchVerticesExecutor.cpp]: maybe else need continue or return, the getOutputSchema can not get it?

File: src/graph/TraverseExecutor.cpp [hunk 19]
```diff
@@ -44,5 +54,74 @@ TraverseExecutor::makeTraverseExecutor(Sentence *sentence, ExecutionContext *ect
...
 }
 
+VariantType Collector::collect(
+        const std::string &prop,
+        const RowReader *reader,
+        RowWriter *writer) const {
+    DCHECK_NOTNULL(reader);
+    VariantType var = getProp(prop, reader);
+    switch (var.which()) {
+        case 0:
```
Comment (wadeliuyi) [inline src/graph/TraverseExecutor.cpp]: a lot place use 0-4, why not define them in enum?

File: src/graph/FetchExecutor.cpp (hunk 15 above)
Comment (CPWstatic) [reply src/graph/FetchExecutor.cpp]: We need a single pr to do this, find too much like this out there.

File: src/graph/FetchEdgesExecutor.cpp [hunk 20]
```diff
@@ -0,0 +1,334 @@
...
+        }
+
+        auto *label = sentence_->label();
+        auto aliasProps = expCtx_->aliasProps();
+        for (auto pair : aliasProps) {
+            if (pair.first != *label) {
+                status = Status::SyntaxError(
+                    "[%s.%s] not declared in %s.",
+                    pair.first.c_str(), pair.second.c_str(), (*label).c_str());
+                break;
```
Comment (laura-ding) [inline src/graph/FetchEdgesExecutor.cpp]: return

File: src/graph/FetchVerticesExecutor.cpp [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+        }
+
+        auto *tag = sentence_->tag();
+        auto aliasProps = expCtx_->aliasProps();
+        for (auto pair : aliasProps) {
+            if (pair.first != *tag) {
+                status = Status::SyntaxError(
+                    "[%s.%s] tag not declared in %s.",
+                    pair.first.c_str(), pair.second.c_str(), (*tag).c_str());
+                break;
```
Comment (laura-ding) [inline src/graph/FetchVerticesExecutor.cpp]: return

File: src/graph/FetchVerticesExecutor.cpp [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+
+void FetchVerticesExecutor::execute() {
+    FLOG_INFO("Executing FetchVertices: %s", sentence_->toString().c_str());
+    auto status = setupVids();
+    if (!status.ok()) {
+        onError_(std::move(status));
+        return;
+    }
+    if (vids_.empty()) {
+        onEmptyInputs();
```
Comment (laura-ding) [inline src/graph/FetchVerticesExecutor.cpp]: ditto

File: src/graph/FetchEdgesExecutor.cpp [hunk 23]
```diff
@@ -0,0 +1,334 @@
...
+    FLOG_INFO("Executing FetchEdges: %s", sentence_->toString().c_str());
+    auto status = setupEdgeKeys();
+    if (!status.ok()) {
+        DCHECK(onError_);
+        onError_(std::move(status));
+        return;
+    }
+
+    if (edgeKeys_.empty()) {
+        onEmptyInputs();
```
Comment (laura-ding) [inline src/graph/FetchEdgesExecutor.cpp]: Should be return?

File: src/parser/TraverseSentences.h [hunk 24]
```diff
@@ -263,6 +263,205 @@ class OrderBySentence final : public Sentence {
...
+    int64_t rank() {
+        return rank_;
+    }
+
+    std::string toString() const;
+
+private:
+    std::unique_ptr<Expression>     srcid_;
+    std::unique_ptr<Expression>     dstid_;
+    int64_t                         rank_;
```
Comment (laura-ding) [inline src/parser/TraverseSentences.h]: ditto

File: src/graph/test/FetchEdgesTest.cpp (hunk 3 above)
Comment (CPWstatic) [reply src/graph/test/FetchEdgesTest.cpp]: _dedup_ is not supported yet .

**Comment by dutor** [comment]
excellent!

Please resolve the conflicts.

File: src/graph/test/FetchEdgesTest.cpp (hunk 1 above)
Comment (CPWstatic) [reply src/graph/test/FetchEdgesTest.cpp]: I'd like to support fetch tag/edge names in the future, so i think fetch prop is OK.

File: src/graph/FetchVerticesExecutor.cpp [hunk 25]
```diff
@@ -0,0 +1,251 @@
...
```
Comment (laura-ding) [inline src/graph/FetchVerticesExecutor.cpp]: why not add return?

File: src/graph/FetchVerticesExecutor.cpp [hunk 26]
```diff
@@ -0,0 +1,251 @@
...
+        return;
+    };
+    auto error = [this] (auto &&e) {
+        LOG(ERROR) << "Exception caught: " << e.what();
+        onError_(Status::Error("Internal error"));
+    };
+    std::move(future).via(runner).thenValue(cb).thenError(error);
+}
+
+StatusOr<std::vector<storage::cpp2::PropDef>> FetchVerticesExecutor::getPropNames() {
```
Comment (laura-ding) [inline src/graph/FetchVerticesExecutor.cpp]: why return use StatusOr?

File: src/graph/FetchEdgesExecutor.h [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+
+
+    using EdgeKeyHashSet = std::unordered_set<
+            storage::cpp2::EdgeKey,
+            std::function<size_t(const storage::cpp2::EdgeKey& key)>>;
+
+private:
+    FetchEdgesSentence                         *sentence_{nullptr};
+    std::vector<storage::cpp2::EdgeKey>         edgeKeys_;
+    EdgeType                                    edgeType_;
```
Comment (laura-ding) [inline src/graph/FetchEdgesExecutor.h]: init

File: src/graph/FetchEdgesExecutor.cpp [hunk 28]
```diff
@@ -0,0 +1,308 @@
...
+        return;
+    };
+    auto error = [this] (auto &&e) {
+        LOG(ERROR) << "Exception caught: " << e.what();
+        onError_(Status::Error("Internal error"));
+    };
+    std::move(future).via(runner).thenValue(cb).thenError(error);
+}
+
+StatusOr<std::vector<storage::cpp2::PropDef>> FetchEdgesExecutor::getPropNames() {
```
Comment (laura-ding) [inline src/graph/FetchEdgesExecutor.cpp]: ditto

File: src/graph/FetchEdgesExecutor.cpp [hunk 29]
```diff
@@ -0,0 +1,308 @@
...
+    auto status = setupEdgeKeys();
+    if (!status.ok()) {
+        DCHECK(onError_);
+        onError_(std::move(status));
+        return;
+    }
+
+    if (edgeKeys_.empty()) {
+        onEmptyInputs();
+    }
```
Comment (laura-ding) [inline src/graph/FetchEdgesExecutor.cpp]: ditto

**Pull request merged**